tests/MongoDB.Driver.Legacy.Tests/MongoClientTests.cs

```diff
@@ -56,5 +60,99 @@ public void UsesSameMongoServerWhenReadPreferenceTagsAreTheSame()
+        public void AreSessionsSupported_should_return_expected_result(string clusterDescriptionJson, bool? expectedResult)
+        {
+            var subject = new MongoClient("mongodb://localhost");
+            var clusterDescription = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+
+            var result = subject.AreSessionsSupported(clusterDescription);
+
+            result.Should().Be(expectedResult);
+        }
+    }
+
+    public class AreSessionsSupportedServerSelectorTests
+    {
+        [Theory]
+        [InlineData("{ clusterType : 'Standalone', servers : [ { state : 'Disconnected', type : 'Unknown' } ]}")]
+        public void SelectServers_should_set_ClusterDescription(string clusterDescriptionJson)
+        {
+            var subject = CreateSubject();
+            var cluster = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+            var connectedServers = cluster.Servers.Where(s => s.State == ServerState.Connected);
+
+            var result = subject.SelectServers(cluster, connectedServers);
+
+            AreSessionsSupportedServerSelectorReflector.ClusterDescription(subject).Should().BeSameAs(cluster);
+        }
+
+        [Theory]
+        [InlineData("{ connectionMode : 'Direct', clusterType : 'ReplicaSet', servers : [ { state : 'Connected', type : 'ReplicaSetArbiter' } ]}")]
+        public void SelectServers_should_return_all_servers_when_connection_mode_is_direct(string clusterDescriptionJson)
+        {
+            var subject = CreateSubject();
+            var cluster = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+            var connectedServers = cluster.Servers.Where(s => s.State == ServerState.Connected).ToList();
+
+            var result = subject.SelectServers(cluster, connectedServers);
+
+            result.Should().BeSameAs(connectedServers);
+        }
+
+        [Theory]
+        [InlineData("{ clusterType : 'ReplicaSet', servers : [ { state : 'Connected', type : 'ReplicaSetArbiter' }, { state : 'Connected', type : 'ReplicaSetPrimary' } ]}")]
+        public void SelectServers_should_return_data_bearing_servers_when_connection_mode_is__not_direct(string clusterDescriptionJson)
+        {
+            var subject = CreateSubject();
+            var cluster = ClusterDescriptionParser.Parse(clusterDescriptionJson);
+            var connectedServers = cluster.Servers.Where(s => s.State == ServerState.Connected).ToList();
+            var dataBearingServers = connectedServers.Skip(1).Take(1);
+
+            var result = subject.SelectServers(cluster, connectedServers);
+
+            result.Should().Equal(dataBearingServers);
+        }
+
+        // private methods
+        private IServerSelector CreateSubject()
+        {
+            return AreSessionsSupportedServerSelectorReflector.CreateInstance();
```

Comment:
We have to use reflection to create the test subject because the class itself is a private nested class of MongoClient.